### PR TITLE
importinto: add metrics and dashboard

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #46704

Problem Summary:

### What is changed and how it works?

add metrics and dashboard for import into tasks to help monitor and diagnose.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

some board last longer than others as it's calculating rate in a time window

img_v2_3bb48b98-d023-442f-b4af-31741eafa8fg
img_v2_1030fc10-0fc8-4a50-9e1b-7ebcd8c0f3cg

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/metric/metric.go [hunk 1]
```diff
@@ -53,200 +53,272 @@ const (
...
+	RowReadSecondsHistogram prometheus.Histogram
+	// RowEncodeSecondsHistogram records the time spent on encoding a batch of rows.
+	RowEncodeSecondsHistogram    prometheus.Histogram
+	BlockDeliverSecondsHistogram prometheus.Histogram
+	BlockDeliverBytesHistogram   *prometheus.HistogramVec
+	BlockDeliverKVPairsHistogram *prometheus.HistogramVec
+}
+
+// NewCommon returns common metrics instance.
+func NewCommon(factory promutil.Factory, namespace, subsystem string, constLabels prometheus.Labels) *Common {
```
Comment (D3Hunter) [inline br/pkg/lightning/metric/metric.go:86]: i am using const labels explicitly, factory might not be needed. @okJiang

File: br/pkg/lightning/metric/metric.go [hunk 2]
```diff
@@ -53,200 +53,272 @@ const (
...
-	LocalStorageUsageBytesGauge          *prometheus.GaugeVec
-	ProgressGauge                        *prometheus.GaugeVec
+	// RowReadBytesHistogram records the number of bytes read from data source
+	// for a batch of rows.
+	// it's a little duplicate with RowsCounter of state = "restored".
+	RowReadBytesHistogram       prometheus.Histogram
+	ChecksumSecondsHistogram    prometheus.Histogram
+	SSTSecondsHistogram         *prometheus.HistogramVec
+	LocalStorageUsageBytesGauge *prometheus.GaugeVec
+	ProgressGauge               *prometheus.GaugeVec
```
Comment (okJiang) [inline br/pkg/lightning/metric/metric.go:215]: Why are some metrics not necessary to import into?

File: br/pkg/lightning/metric/metric.go [hunk 3]
```diff
@@ -53,200 +53,272 @@ const (
...
 )
 
+// Common contains metrics shared for lightning and import-into.
+// they will have different namespace.
+// metrics here will an additional task-id const-label when used for import-into.
+type Common struct {
+	ChunkCounter *prometheus.CounterVec
+	// BytesCounter records the total bytes processed.
+	// it has a state label, values includes:
+	// 	- total_restore: total source file bytes needs to restore, it's constant.
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply br/pkg/lightning/metric/metric.go:67]: that's how it's named in lightning, we don't change it here

File: br/pkg/lightning/metric/metric.go [hunk 4]
```diff
@@ -53,200 +53,272 @@ const (
...
+	// 	- restored: source file bytes encoded and sorted.
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply br/pkg/lightning/metric/metric.go:68]: in lightning, we use `restore` to represent `encode & sort`. will explain more to it

File: br/pkg/lightning/metric/metric.go (hunk 2 above)
Comment (D3Hunter) [reply br/pkg/lightning/metric/metric.go:215]: some metric is added in lightning, but not used in dashboard. in this pr only add those needed in the panel

and some're not fit for import into, as we only import 1 table here, such as ChecksumSecondsHistogram

File: br/pkg/lightning/metric/metric.go [hunk 5]
```diff
@@ -378,17 +436,33 @@ func ReadHistogramSum(histogram prometheus.Histogram) float64 {
 	return metric.Histogram.GetSampleSum()
 }
 
-type ctxKeyType struct{}
+type ctxKeyType string
 
-var ctxKey ctxKeyType
+var (
+	allMetricKey    ctxKeyType = "all-metrics"
+	commonMetricKey ctxKeyType = "common-metrics"
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply br/pkg/lightning/metric/metric.go:443]: it's very similar to `common`. let's wait other reviewer's opinions

**Review by okJiang — APPROVED** [review]
rest lgtm

File: disttask/importinto/dispatcher.go [hunk 6]
```diff
@@ -220,6 +221,10 @@ func (dsp *importDispatcherExt) OnNextStage(ctx context.Context, handle dispatch
 	var nextStep int64
 	switch gTask.Step {
 	case proto.StepInit:
+		if metrics, ok := metric.GetCommonMetric(ctx); ok {
```
Comment (ywqzzy) [inline disttask/importinto/dispatcher.go:224]: Why not add this counter when generating the taskMeta.Plan?

**Review by ywqzzy — APPROVED** [review]

File: disttask/importinto/dispatcher.go (hunk 6 above)
Comment (D3Hunter) [reply disttask/importinto/dispatcher.go:224]: you mean on creating task? we only want this metric after the task start running

**Pull request merged**